### PR TITLE
 Add drop_columns to dataframe api

### LINKED ISSUE #11007: Add drop_columns to dataframe api
### Is your feature request related to a problem or challenge?

The dataframe api has select_columns but does not have drop_columns. Having a drop_columns function would add ease of use to the datafusion api and would mirror other common dataframe apis that have this functionality

- https://docs.rs/polars/latest/polars/frame/struct.DataFrame.html#method.drop
- https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#drop(col:org.apache.spark.sql.Column,cols:org.apache.spark.sql.Column*):org.apache.spark.sql.DataFrame
- https://docs.snowflake.com/developer-guide/snowpark/reference/scala/com/snowflake/snowpark/DataFrame.html#drop(cols:Array[com.snowflake.snowpark.Column]):com.snowflake.snowpark.DataFrame

### Describe the solution you'd like

Anew function  `drop_columns(self, columns: &[&str]) -> Result<DataFrame>` added to the dataframe api

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #11007 

## Rationale for this change

Add drop_columns to the dataframe api.

## What changes are included in this PR?

code, tests, documentation

## Are these changes tested?
Yes via tests in dataframe/mod.rs

## Are there any user-facing changes?

dataframe API was amended.

### CODE REVIEW

File: datafusion/core/src/dataframe/mod.rs [hunk 1]
```diff
@@ -244,6 +244,41 @@ impl DataFrame {
...
+    /// ```
+    pub fn drop_columns(self, columns: &[&str]) -> Result<DataFrame> {
+        let fields_to_drop = columns
+            .iter()
+            .map(|name| {
+                self.plan
+                    .schema()
+                    .qualified_field_with_unqualified_name(name)
+            })
+            .collect::<Result<Vec<_>>>()?;
```
Comment (waynexia) [inline datafusion/core/src/dataframe/mod.rs:268]: collect into a hash table might be good for wide table scenarios

**Review by waynexia — APPROVED** [review]
LGTM, thanks @Omega359

I left a little suggestion about the container's type.

**Review by comphead — COMMENTED** [review]
Thanks @Omega359 
Appreciate if you add more tests, like drop with empty array
Drop where columns are with double quotes, drop non existent cols, drop duplicated

File: datafusion/core/src/dataframe/mod.rs (hunk 1 above)
Comment (Omega359) [reply datafusion/core/src/dataframe/mod.rs:268]: Interesting, I'll take a look. The logic for this function was based on the select_columns fn which did things that way.

**Comment by alamb** [comment]
Thanks @Omega359  @comphead  and @waynexia  -- the new tests look good to me 🚀

**Pull request merged**